wandb/sdk/wandb_run.py

```diff
@@ -1469,47 +1469,68 @@ def _datatypes_callback(self, fname: str) -> None:
         files: FilesDict = dict(files=[(GlobStr(glob.escape(fname)), "now")])
         self._backend.interface.publish_files(files)
 
-    def _visualization_hack(self, row: dict[str, Any]) -> dict[str, Any]:
+    def _visualization_hack(
+        self,
+        data: dict[str, Any],
+    ) -> dict[str, Any]:
+        """Transforms visualization objects into table objects for a given dictionary of data.
+
+        This function handles special visualization types (Visualize and CustomChart) by:
+        - For Visualize objects: Replaces them with their underlying table
+        - For CustomChart objects:
+           - Replaces the CustomChart object with its table in the log history
+           - Modifies the config to use the table key when adding the graphgraphing
+
+        Args:
+            data: Dictionary containing logged data that may include visualization objects
+                Visualization objects can be nested in dictionaries, which will be processed recursively.
+
+        Returns:
+            The processed dictionary with visualizations handled transformed into tables.
+        """
         # TODO(jhr): move visualize hack somewhere else
-        chart_keys = set()
-        for k, v in row.items():
-            if isinstance(v, Visualize):
-                row[k] = v.table
+        keys_to_remove = set()
+        for k, v in data.items():
+            if isinstance(v, (Visualize, CustomChart)):
+                if isinstance(v, Visualize):
+                    data[k] = v.table
+                # If this is a custom chart, we will update our history with the table key.
+                # Allowing for charts to be nested in dictionaries.
+                elif isinstance(v, CustomChart):
+                    keys_to_remove.add(k)
+
                 v.set_key(k)
                 self._config_callback(
                     val=v.spec.config_value,
                     key=v.spec.config_key,
                 )
```

Comment:
```suggestion
                if isinstance(v, Visualize):
                    data[k] = v.table
                # If this is a custom chart, we will update our history with the table key.
                # Allowing for charts to be nested in dictionaries.
                elif isinstance(v, CustomChart):
                    keys_to_remove.add(k)

                v.set_key(k)
                self._config_callback(
                    val=v.spec.config_value,
                    key=v.spec.config_key,
                )
```
the outer if is redundant

Reply:
I personally felt like doing two layers was a little more clean.

Because for both `Visualize` and `CustomChart` we need to do 
```
v.set_key(k)
self._config_callback(
    val=v.spec.config_value,
    key=v.spec.config_key,
)
```

in both cases, and this allows us to only have to do it once. Where as before this change it was copy pasted for both cases.

Reply:
more level of nesting is not cleaner as a rule, also, this way it is obvious what you are doing:
```
if isinstance(v, Visualize):
    process_visualize(...)
elif isinstance(v, CustomChart):
    process_charts(...)
elif isinstance(v, dict):
    ...
```
where `process_visualize` and `process_charts` could be completely different!